### ISSUE TITLE
Allow "underscore" in username?

### ISSUE THREAD

**Issue by gt-tech** [isaacs/github#906]
As per current github user rules, username can only have alphanumeric characters or a hyphen. 

This turns out to be an issue when:

1. User wants to produce a java based open-source product hosted from GITHUB repo
2. User wants to use his github qualified namespace as groupId and hence by convention, his top level java package would be something like: `org.github.<username>`
3. User's choice of alphabetic username isn't available (taken by someone else) and he gets forced to use john**-**doe instead of john**_**doe (note that having numbers in username to make it unique would make the java package declaration even worse)
4. john-doe now can't be used as a package name for user's source code as Java doesn't rightfully allow "-" in package names.

Hence having "_" in username as allowed character appears to be a nice feature unless there's a technical reason blocking it to happen on github.

**Comment by cirosantilli** [comment]
converting `-` to `_` looks like a common java naming convention: http://stackoverflow.com/questions/3179216/what-is-the-convention-for-word-separator-in-java-package-names

`john-doe` vs `john_doe` would increase confusion I think (yes, there is also `johndoe`)

**Comment by gt-tech** [comment]
Good point but I don't wish to debate on whether john_doe makes sense for comparing against johndoe as two separate namespaces, point well taken there's a branding decision that user has to make but i vote for this enhancement because java allows for it, DNS allows for it so unless there's a tech reason for github to not support it, i think it should. github probably wouldn't want to impose a strong opinionated thought about that someone shouldn't take john_doe if johndoe is take up someone else, that's a user choice in my view.

**Comment by smcv** [comment]
Github seems very unlikely to allow underscores, because DNS labels containing underscores are not valid Internet hostnames (reference: [RFC 1123 "Requirements for Internet Hosts — Application and Support" §2.1](https://tools.ietf.org/html/rfc1123#page-13)) so Github would be unable to allocate `john_doe.github.io` and similar names. Yes, the DNS protocol supports labels that contain underscores, but in the contexts where they're used in practice (such as DNS SRV), they are used precisely *because* they are not valid in Internet hostnames, to avoid colliding with an ordinary hostname. Github also doesn't allow usernames that start with a dash, which is also an Internet hostname restriction.

Assuming github continues to only allow dashes and not underscores, in domain-name-derived contexts that only accept underscores (like D-Bus interfaces and Java packages) you can replace dashes with underscores and know that you won't cause a collision: `com.github.john_doe.AwesomeApp` can only belong to Github user `john-doe`, because `john_doe` can't exist. If both underscores and dashes were allowed, that reasoning would not be valid.

The standard advice that the D-Bus maintainers give for domain-name-based naming is to replace dashes with underscores, and prepend an underscore to any component that starts with a digit (which also isn't allowed, the same as in Java): so if 7-zip.org (which hits both of the problem cases) wanted to publish an archiver that communicated via D-Bus, they should call it something like `org._7_zip.Archiver`.

**Comment by ap0phi5** [comment]
GitHub uses underscores in its usernames already, but for its "Enterprise Managed Users" feature (part of using SSO). It's a reserved special character used internally, so can't be used by the general public.
